### ISSUE TITLE
ViewComponents as Widgets / Handling GET vs POST

### ISSUE THREAD

**Issue by Antaris** [aspnet/Mvc#2798]
Preface: This is not a bug or an issue, just an enquiry.

Previous, on MVC5, we've created a Component/Widget framework around the use of MVC child actions. 

Essentially, in my MVC5 view, I can do the following:

```
@Component("<component-name>, <component-data>);
```

An example could be `@Component("ContactUs-Form", new { recipient = "yada@somewhere.com" })`.

This would internally call `@Html.RouteUrl` and allow us to execute a child HTTP request as a child action. An example controller might be:

```
public class ContactFormComponentController : ComponentController
{
    public ActionResult Form(string recipient = null)
    {
        return View("ContactForm", new ContactFormViewModel { Recipient = recipient });
    }

    [HttpPost]
    public ActionResult Form(ContactFormViewModel model)
    {
        if (ModelState.IsValid)
        {
            // Do something
            return View("ContactFormSubmitted", model);
        }

        return View("ContactForm", model);
    }
}
```

This is a very naive implementation, but the import thing is, because it is a controller, follows the same rules for action resolution. If the page that embeds the component is a GET, the form is displayed, if it is posted back, the component will run the POST method. We do a few other things to ensure that only the correct component processes an appropriate POST method if there are multiple components on the page. The important thing is, we can drop the component on any page and it should work.

Now in MVC6, we have `ViewComponent` which is the successor to child actions, but the API for a ViewComponent is a little more streamlined:

```
public class ContactFormViewComponent : ViewComponent
{
    public IViewComponentResult Invoke(string recipient = null)
    {
        //
    }
}
```

With this in mind, how could we approach a similar architecture with view components, whereby a component can handle GET and POST (and possibly multiple POSTs - think multi-state components), or would we stick with child actions (are child actions still supported?). Do view components support model binding for arguments also?

This is not a general use case, but more for framework devs building things like CMS on MVC6.

**Comment by marcuslindblom** [comment]
:+1:

**Comment by PromontoryProtean** [comment]
I am interested to hear some guidance on this subject as well. I am beginning to explorer porting a WebForms based CMS to asp.net vnext, and I am running into similar roadblocks. Model binding of some kind in reusable components would be extremely useful. I really like the concepts introduced with view components, and I understand how fundamentally different MVC is from WebForms. But I also think that a practical mechanism to transfer postback data into view components (or something similar) would add tremendous value for CMS developers.

**Comment by rubenprins** [comment]
:+1:

**Comment by ctolkien** [comment]
>But I also think that a practical mechanism to transfer postback data into view components (or something similar) would add tremendous value for CMS developers.

You still have access to the httpcontext which has the request method (post or get, for instance) any any posted form data.

So it's doable currently, it's just not as "clean" as perhaps having different components getting involved for different http verbs.

**Comment by danroth27** [comment]
Related to #2338.

**Comment by toddtsic** [comment]
Not sure if this directly relates:

I'm building a master->detail view with master being search results of registrants, and detail being their details, am conceiving the details view as a view component (as I would a user control in webforms).  My question is how to refresh/update the details view (initial call from @await Component.InvokeAsync("SearchRegistrantDetails", "RegistrationID")

Is it possible to re-invoke the view component?

**Comment by rynowak** [comment]
Thanks for all of the feedback here. We're going to hurry up with #2338 - which makes this easier. Right now there's a lot of boilerplate required. I'll post a sample when I have it done so anyone using an earlier build than beta-7 can start using this.

We're not planning to have support for child actions or any kind of 'simulated request' model for this release. The ViewComponent programming model splits the responsibilities that used to be a part of child-actions in a way that has less overlap with regular controllers/actions. Now only controller/actions are exposed externally, and only controller/actions perform model binding or have filters. 

If you want to support a 'widget' type model using a VC, I'd recommend exposing a controller endpoint that performs the necessary logic invokes the view component. This provides a separation between things that apply to HTTP (routing, filters, model binding, validation) and things that don't (rendering, retrieving data). We need to do some work to make this easy (#2338) as right now as the code required is pretty ugly. 

For anyone porting widget-style child actions to the new framework, we're interested in knowing what your pain points are, and what could be made easier.

**Comment by Antaris** [comment]
@rynowak @danroth27 I've pushed a very basic/naive example of what we can currently achieve through child actions. [MVCWidget](https://github.com/Antaris/MVCWidget)

This is just a basic ASP.NET MVC 5 project, but the main parts are as follows:

- Created a controller, the `ContactUsController` which implements two overloaded methods called `Form`, one for `GET` one for `POST`. The post method takes in an instance of `ContactUsViewModel`. 
- Added a HTML helper extension method called `Component` passing in a component name. In this simple example, I've merely mapped this to the controller, and used a simple switch to determine the action name. In a fuller framework, this would probably go off to some `IComponentProvider` of sorts, be it database or code-based, etc. 

The helper method merely wraps to `Html.Action`, but allows us to pass in some context information, perhaps the `Page` model of the outer page. When the page runs in reponse to a `GET` request, the `ContactUsController` merely returns the partial view. In response to the `POST` request, the model binder kicks in and binds the model and highlights any validation issues so we can determine what we are doing with the submitted data. In this sense, components get the same model binding abilities that full controllers do, and the same action resolutions also, because they are essentially components, we just orchestrate them a specific way.

If you clone the repo, and run the Contact page of the MVC project, you'll see how it is plumbed together. Obviously this is a very simple implementation, and doesn't consider the following features which we've already implemented to complete the Widget experience:

- Enforcing only unique widgets are processed on the page, typically with a `[FormValueRequired]` or a custom `[WidgetResult("ContactUs")]` attribute.
- Handling when a child request wants to perform a redirect - consider a login Widget. We handle this by dumping the `RedirectResult` into `HttpContext.Items` and handling it when we're back up at the parent request.
- Multi-step form components.

In my developed framework, we don't actually use `Html.Action`, we manually gear up the child action code ourselves so we can handle the pre-execute (to prepare all the route data) and post-execute (to handle specific Widget responses, like redirects).

Looking over your comments, particularly `If you want to support a 'widget' type model using a VC, I'd recommend exposing a controller endpoint`, that in itself assumes either a) child actions which you no longer support, or b) AJAX-based requests just for the view component itself, which means we can't process the widget as part of the page. Having a controller return a `Component` in MVC 5 is useful, but it still doesn't afford us the following features:

- Model binding
- Action resolution based on HTTP verbs.

I am wondering if there are two possible alternatives, either:

1. Implement components as Tag Helpers somehow
2. Implement components as View Components using a subclass of `ViewComponent` which performs some sort of model binding and action resolution itself as part of its base `Invoke`/`InvokeAsync`?

**Comment by PromontoryProtean** [comment]
Rynowak, I'm trying to wrap my head around all of this, but I will list the pain points that I have hit after doing some prototyping.

First I would like to see an overload to allow calling a View Component using a file path, similar to what we can do with full and partial views. I feel as though loading View Components views entirely by location convention is limiting because when building a pluggable widget or component framework we will need a way to organize component hierarchies in different ways. We can already put View Component classes wherever we want. Why not give us the flexibility to organize View Component views into a file hierarchy that can match whatever folder structure that a particular CMS framework might require?

Next, I would like to be able to load View Components programmatically, not only in a controller but also from procs such as helper methods located in external assemblies. In my scenario I attempted to created a "Content Zone" Tag Helper that would load View Components dynamically based on a list pushed in from the view model, but I could not figure out how to call the View Component. I attempted to inject IViewComponentHelper using constructor injection and calling ViewComponentHelper.Invoke() in my Tag Helper but that did not work. Is this a scenario that that #2338 will potentially address?

Next I would love to have some mechanism to do model binding. It looks like this is going to be impractical based on my understanding of the architecture, but it is still on my wish list. I'll have to investigate your suggestion of using controller endpoints further as I am relatively new to MVC.

I have a long history of doing WebForms development, but I am trying to be pragmatic and stay away from asking for WebForms concepts to be implemented in MVC. Hopefully I am not sounding too much like a WebForms control builder.

**Comment by rynowak** [comment]
I want to be straightforward about the fact that we as a team haven't done much yet to support building 3rd party frameworks, which it seems like both of you are doing. I want to try and help with any and all of your feedback, but don't be surprised if you find ragged edges trying to extend the system.

A few clarifying comments about the design and intent of ViewComponents, before I try to address your more specific points.

- View Components are a replacement for child actions. There are no child actions in MVC6.

- View Components are simplified from child actions. They are not conceptually equivalent and don't do all of the same things in the same ways.

- View Components are not reachable directly as an HTTP endpoint, they are invoked from your code (usually in a view). This also means that a View Component never 'handles' a request.

- View Components do not perform model binding. They can have parameters which get the values passed in by your code.

**Comment by rynowak** [comment]
@Antaris - responding to your comments here

>Looking over your comments, particularly If you want to support a 'widget' type model using a VC, I'd recommend exposing a controller endpoint, that in itself assumes either a) child actions which you no longer support, or b) AJAX-based requests just for the view component itself, which means we can't process the widget as part of the page. 

**a.** is correct, there are no child actions

>Model binding
>Action resolution based on HTTP verbs.

Processing an HTTP request involves both understanding user input (model binding & validation) and crafting a *full* response. View Components are intended to be isolated from these concerns, and handle a **part** of the rendering logic for a response. The intent is to make them more simplified and reusable.

In some cases, you want to be able to response to a request with just a plain 200 and the output of a VC. That's the case for #2338. We want this to be explicit on the part of the author rather than a VC *automatically* being promoted to an HTTP endpoint.

> Implement components as Tag Helpers somehow

We've built samples for this, and have an item sitting on the backlog for ideas around the theme of "making View Components easier to invoke". #1051 #1435 If you have more suggestions or just want to comment on what's there then feel free.

>Implement components as View Components using a subclass of ViewComponent which performs some sort of model binding and action resolution itself as part of its base Invoke/InvokeAsync?

The intent of View Components is that the data needed is passed in by the calling code (usually a view). Having seen the rest of my responses, can you explain a bit more about what you would use this for?

**Comment by rynowak** [comment]
@PromontoryProtean - responding to your comments here

> First I would like to see an overload to allow calling a View Component using a file path, similar to what we can do with full and partial views. 

Can you explain what you mean? (Maybe an example). Do you mean returning a view from a VC **or** the code calling into the VC?

> I feel as though loading View Components views entirely by location convention is limiting because when building a pluggable widget or component framework we will need a way to organize component hierarchies in different ways.

If you're building a framework you want to go even further with this. You probably want to replace the view engine so that you can serve embedded precompiled content. Example here: https://github.com/aspnet/Mvc/tree/dev/test/WebSites/RazorEmbeddedViewsWebSite


> Next, I would like to be able to load View Components programmatically, not only in a controller but also from procs such as helper methods located in external assemblies. 

View Components expect to be called in the context of an action, so you'll need to have an `ActionContext` or do a convincing job faking it. 

> I attempted to inject IViewComponentHelper using constructor injection and calling ViewComponentHelper.Invoke() in my Tag Helper but that did not work. Is this a scenario that that #2338 will potentially address?

#2338 will solve a different scenario, but the code there will show you what you need to do to use `IViewComponentHelper`. You need to call the `Contextualize(...)` method. Anything that implements `ICanHasViewContext` requires an extra initialization step right now.

> Next I would love to have some mechanism to do model binding in View Components. It looks like this is going to be impractical based on my understanding of the architecture, but it is still on my wish list. I'll have to investigate your suggestion of using controller endpoints further as I am relatively new to MVC.

Try the controller approach and let us know once #2338 is done. I'm of the mind that having to explicitly define an action as an endpoint is a good thing. If you need to add filters or any other features, you have a place to put them. If we did something to "promote" VCs to actions, it would be a dead-end for customization.

**Comment by Antaris** [comment]
@rynowak Thanks for getting back to us, feedback follows

> The intent of View Components is that the data needed is passed in by the calling code (usually a view). Having seen the rest of my responses, can you explain a bit more about what you would use this for?

Yeah, maybe we're crossing wires a bit here. I am very much **not** asking you guys to implement widgets, just looking at how *we as framework developers* could enable these scenarios in our frameworks built on top of ASP.NET 5.

In my view, what a widget is, is a self contained display unit, which could posses one of more states. A widget can be injected into any page, and it should be entirely self contained. That is not to say that it can't receive arguments from the parent page, but generally it should enable scenarios where it can be injected into dissimilar context pages. e.g., I am using a vNext powered CMS, and that CMS allows me to create my homepage. Now on my homepage, I want to insert a widget which allows me to list my latest blog articles. In this scenario we can probably acheive that using a `ViewComponent` directly.

Now a more complex scenario could be say, I've developed an enquiry component (which is a form with two states, the initial form state, and the success state). On all my product pages, I want to inject my enquiry component, so they can launch an enquiry from all my product pages. This enquiry component should be able to pull some contextual information from the page (i.e., what product am I for), but then also pull information from a source (e.g., what countries to populate in a form field). In this instance, this is a `<form>` based component, so it needs to post *somewhere*. If it were possible to `POST` back to the same page, the widget which is in place on the page can then tackle the `POST` state.

In MVC5 this is already possible as child actions, and because widgets are implemented as full controllers, they get the full benefits from model binding and validation, which is *required* as part of the `POST` state.

Perhaps ViewComponents are not the right model for this in vNext, but I don't see any other alternative yet. I understand we're still at beta5/6 stage so perhaps things can be introduced/changed that makes this easier. 

From what I see it, you have two main developer stories with MVC, you have the developer who is building on top of MVC to make a website, and you have the developer who is building a platform/framework on top of MVC. The latter may not be the target for this initial release, but I do fear that if these sorts of scenarios are not considered, would it mean a fair amount of refactoring when these scenarios are eventually implemented, who knows.

As an aside, and a potential way to tackle this, is it possible to manually trigger model binding and validation, from within a `ViewComponent`, so that we could do something similar to:

    public class ContactFormViewComponent : ViewComponent
    {
      private readonly IContactFormService _service;
      private readonly IHttpContextAccessor _accessor;
      
      public ContactFormViewComponent([NotNull] IContactFormService service, [NotNull] IHttpContextAccessorr accessor)
      {
        _service = service;
        _accessor = accessor;
      }
      
      public IViewComponentResult Invoke([NotNUll] PageViewModel model)
      {
        var httpContext = _accessor.HttpContext;
        if (httpContext.Request.Method == "POST") 
        {
          return Post(BindModel<ContactFormViewModel>(), model);
        }
        
        return Get(new ContactFormViewModel { Page = model });
      }
      
      private IViewComponentResult Get(ContactFormViewModel model)
      {
        return View("Get", model);
      }
      
      private Task<IViewComponentResult> Post(ModelBindResult<ContactUsViewModel> result, PageViewModel model)
      {
        result.Model.Page = model;
      
        if (result.ModelState.IsValid)
        {
          _service.Save(result.Model);
          return View("Post", result.Model);
        }
        else 
        {
          return Get(result.Model);
        }
      }
    }

Obviously if this type of scenario was possible with view components, there is still a fair amount of grunt work to do. With the current implementation, if we wanted to support these postback scenarios, the suggestion I guess would be to have a `ViewComponent` that handles the initial display, and then a seperate `Controller` to handle the postback, with a redirect back to the source page. This poses a few problems:

1. The logic for the 'widget' is implemented in three places, the view component, the controller and the view, making it harder to maintain and test.
2. We probably can't give a decent feedback mechanism, if validation fails, as we have to redirect back to the source page. We could stuff something in temp data, but it makes it difficult to figure out which component (of potentially many) on the page failed.

Again, I'm not trying to be difficult in any way or try and get anybody frustrated, but I'm just trying to bounce ideas to see what solutions we as a community might come up with, to tackle what I think is probably a common development scenario for framework devs; I appreciate this is not likely one of the intended audiences for the initial RTM?

**Comment by marcuslindblom** [comment]
This is exactly what I need for my CMS build, great example with the enquiry form on every page. :+1:

**Comment by davidfowl** [comment]
@Antaris how much of the pipeline do you actually want to run? All of the filters? Just model binding and validation?

**Comment by rubenprins** [comment]
I think (correct me if I'm wrong!) that the main issues and questions here are caused by view components being a refactoring of complex partial views into a view and a class, rather than a single view with `@inject`, extensive `@{ ... }` code blocks and ad hoc model/ViewBag parametrization. Which is a good thing.

But this happens to be the most simplified scenario of what a child action can (could) handle. Which at the moment is also the most advanced scenario a view component *on itself* can handle.

1. VCs do not support client-server interaction ("actions"), at least not without a supporting buddy-controller (not even a refresh/re-render can be implemented purely by a view component).
2. VCs do not support the notion of their own model (having no model binding), therefore there's also no model validation. As far as a VC is concerned there *is* no input model, just arguments.
3. VCs do not have state: either the views that call them need to know how preserve and pass back the VC's state somehow, or a VC needs to invent its own state management, probably by checking for a POST and reading the Form collection inside the Invoke() method (using special naming conventions to prevent clashing with parent views).

From the perspective of someone growing up with web forms: child actions were already fairly fragile and restricted compared to user controls from an isolation and state management point of view. I'm no advocate of the single form/__VIEWSTATE solution though! Far from it. But view components currently make server side componentization of pages even more complicated.

> how much of the pipeline do you actually want to run? All of the filters? Just model binding and validation? 

Probably at the very least a yes to model binding and validation. (With `HtmlNamePrefix` support for multiple VCs per page?)

But being able to "upgrade" a view component to a (mini?) controller would probably make things a lot cleaner when you need to respond to user interaction.

Because when I have a buddy controller responding to user actions, I already have model binding, validation, DI, action dispatching/filtering, business logic, etc. Why do I need a separate view component in the first place? Returning partial views from the buddy controller's actions would be entirely sufficient. And you don't want to spread code across a buddy controller and the view component.

**Comment by PromontoryProtean** [comment]
@rynowak - you asked me to clarify this:
> First I would like to see an overload to allow calling a View Component using a file path, similar to what we can do with full and partial views.

What I mean is that in the View Component's class we return a view using something like:
`return View();`

According to the docs:
>Create a new folder called PriorityList in under the Views\Todo\Components folder. This folder name must match the name of the view component class, or the name of the class minus the suffix (if we followed convention and used the ViewComponent suffix in the class name). If you used the the ViewComponent attribute, the class name would need to match the attribute designation.

We are still held to the convention even if we use a View name by calling the view using the view name overload :
`return View("MyView");`

What I would like to do is specify a full file path in order to ignore the conventions, something such as:
`return View("~/Widgets/MyComponent/MyView.cshtml");`

Unless I am mistaken, we can do this to return full and partial views. I think it would be helpful for framework writers to able to specify the full path using an alternate overload of some kind when returning View Components views.

**Comment by PromontoryProtean** [comment]
@Antaris, you did a great job of explaining the challenges that framework developers are going to run into and I think your proposed solutions deserve careful consideration. @rubenprins also makes great points, especially:

>Probably at the very least a yes to model binding and validation. (With HtmlNamePrefix support for multiple VCs per page?)

My guess is that many developers moving over to vnext from Web Forms are going to be looking for ways to build components that can encapsulate postback functionality. And I think there will eventually be multitudes of developers making the switch, much more so than migrated from Web Forms to earlier versions of MVC.

 That being said, I also think I understand why View Components are designed as they are, and I also understand why a lot of MVC developers might not like some of these ideas. But ultimately I think there is huge value in finding a middle ground here. Adding the ability to build a component (View Component, child action, partial or whatever) that can do model binding and handle validation would be a huge benefit all around.

**Comment by rynowak** [comment]
> is it possible to manually trigger model binding and validation

@Antaris - the methods on `ModelBindingHelper` get you pretty close. 

> What I would like to do is specify a full file path in order to ignore the conventions, something such as: `return View("~/Widgets/MyComponent/MyView.cshtml");`

@PromontoryProtean - it sounds like this is just a bug. It looks the logic for view components and controllers diverge in this case - view components don't property handle the tilde. For now if you just use `/Widgets/MyComponent/MyView.cshtml` it looks like it will work. Filed: #2856

**Comment by PromontoryProtean** [comment]
I have finally had some time to install beta 6 and dig further, so I figured I'd post back with some more feedback.

@rynowak, I was able to get around #2856 by dropping the tilde (I will add the tilde back once beta 7 is available). The following works as expected. 
`return View("/Widgets/MyComponent/MyView.cshtml");`

I was able to accomplish my goal of returning view components within a tag helper by following your suggestion to look at ViewComponentHelper.Contextualize. It does work but it is a little cumbersome and not very easy to discover. I'll take another look at #2338 when beta 7 is released and see if this has changed the process in any way.

I do not think the suggestion of using controller endpoints will work in my framework. I think @Antaris summed it up when he said it assumes either a) child actions (not available) or b) AJAX requests, and I need to be able to process the widgets as part of the page on the server.

As it stands the biggest challenge for me to solve is going to be adding widgets at run time that support postback and databinding. I am currently brain storming and prototyping, but I am definitely open to further suggestions. As it stands there is nothing I can see built into the framework that solves this in a straight forward way.

**Comment by Antaris** [comment]
For any of those still interested, I've built a prototype Widget framework based on the design of ViewComponents. It supports model binder-source arguments, and has a slightly different approach to method selection for invocation. I've purposely not tried to modify ViewComponents to support this model of execution, mainly because the way in which method arguments are provided conflicts. This currently only targets `rc1-final` packages, I haven't targeted bleeding edge.

https://github.com/Antaris/AspNetCore.Mvc.Widgets

Would appreciate feedback. Any issues please raise them against that repo instead of this one.

cc @marcuslindblom @PromontoryProtean @rubenprins @ctolkien @danroth27 @rynowak

**Comment by rynowak** [comment]
That's pretty cool! Is a widget itself exposed as an HTTP endpoint or does it require a page to include it when you do a POST?

> has a slightly different approach to method selection for invocation

We've simplified this substantially in RC2, invoking a VC now takes a dictionary-ish (most likely an anonymous type) to contain the arguments.

**Comment by Antaris** [comment]
@rynowak

I guess you could do both. The current model does all the processing in placement on a page but with a little extra tweak it could probably have an API controller which forwards to the widget by returning a WidgetResult much like you've added a ViewComponentResult.

If RC2 uses a dictionary approach to the invocation arguments then perhapshe it would be better for Widget to subclass ViewComponent and just provide a custom invoker because the infrastructure is largely the same?

**Comment by rynowak** [comment]
> If RC2 uses a dictionary approach to the invocation arguments then perhaps it would be better for Widget to subclass ViewComponent and just provide a custom invoker because the infrastructure is largely the same?

Yeah it seems like your framework could just be a custom invoker in RC2. I'd be happy to help if you find any sticking points or extensibility that needs to be made for first class for this release (open issues).

We like a lot of the ideas here, but we don't want to add more features this late in the release. We'd be happy to help you promote this package so people can find it.

**Comment by Antaris** [comment]
Thanks, appreciate the help and support. I'd like to polish this some more, I've got a few things I'd like to iron out - including making sure the method selection works well enough. I've had a few ideas around multi-state widgets.

I'll look at branching an RC2 version so I can investigate any changes, and I'll raise any issues as appropriate. I'm likely to blog this at some point to, to at least further explain the concepts.

Thanks for your help :-)

**Comment by PromontoryProtean** [comment]
@Antaris, nicely done. I checked out your repository, and it looks like it fills in most, if not all of the blanks. I haven't had time to work with the source yet, but I'll download it soon and try it out in the CMS framework that I was experimenting with. I'll try to provide some useful feedback in your repository.

**Comment by Antaris** [comment]
@rynowak @PromontoryProtean I've now updated my library to be compatible with RC2 plus has some new features:

It is now possible to have the same widget be hosted multiple times on the same page and be able to process individually. To do this, you would need to pass in a unique id for each instance of the widget, e.g.:

```
@await Widget.InvokeAsync("ContactForm", widgetId: "Form1");
@await Widget.InvokeAsync("ContactForm", widgetId: "Form2");
```

And with a new tag helper, I can now generate the correct hidden inputs to wire this all up for you:

```
<form method="post" widget-form>...</form>
```

It is also possible to have multi-state widgets, where each step could be an individual form, e.g.

```
public class WizardWidget : Widget  
{  
  public IWidgetResult InvokeGet()  
  {  
    return View("StepA");  
  }  
 
  public IWidgetResult InvokeStepAPost(WizardViewModel model)  
  {  
    if (ModelState.IsValid)  
    {  
      return View("StepB", model);  
    }  
 
    return View("StepA", model);  
  }  
 
  public IWidgetResult InvokeStepBPost(WizardViewModel model)  
  {  
    if (ModelState.IsValid)  
    {  
      return View("Confirmation", model);  
    }  
 
    return View("StepB");  
  }  
}  
```

 
Which can be paired with a set of views:  
 
```html  
StepA.cshtml:  
<form method="post" widget-state="StepA">  
  ...  
</form>  
 
StepB.cshtml:  
<form method="post" widget-state="StepB">  
  ...  
</form>  
``` 

Widgets can also be surfaced directly from controllers the same way view components can, simply return a `WidgetResult`:

```
public IActionResult GetWidget()
{
   return this.Widget("ContactForm");
}
```

The framework still very closely aligns with view components but there are subtle differences, such as method selection, etc. and support for model binding. 

There is a small bug since migrating to RC2 where the validator doesn't fail a `[Required]`-decorated property on the model - not sure what the issue is yet.

I'd be very much interested in your thoughts in regards to whether this sits within MVC as a first-class feature (future feature), or maybe a Contrib project?

Let me know your thoughts. I haven't pushed this to NuGet yet either.

https://github.com/Antaris/AspNetCore.Mvc.Widgets

**Comment by Eilon** [comment]
We are closing this issue because no further action is planned for this issue. If you still have any issues or questions, please log a new issue with any additional details that you have.
